java-support/src/test/scala/io/cloudstate/javasupport/impl/eventsourced/EventSourcedImplSpec.scala

```diff
@@ -0,0 +1,267 @@
+      entity.send(init(ShoppingCart.Name, "cart1"))
+      entity.send(command(1, "cart2", "command"))
+      entity.expect(failure(1, "Protocol error: Receiving entity is not the intended recipient of command"))
+      entity.expectClosed()
+    }
+
+    "fail when entity is sent multiple init" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.expect(failure("Protocol error: Entity already inited"))
+      entity.expectClosed()
+    }
+
+    "fail when entity is sent empty message" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(EmptyMessage)
+      entity.expect(failure("Protocol error: Received empty/unknown message"))
+      entity.expectClosed()
+    }
+
+    "fail when snapshot handler does not exist" in {
+      val entity = client.connect
+      val notSnapshot = domainLineItem("?", "not a cart snapshot", 1)
+      val snapshotClass = notSnapshot.getClass
+      entity.send(init(ShoppingCart.Name, "cart", snapshot(42, notSnapshot)))
+      entity.expect(failure(s"No snapshot handler found for snapshot $snapshotClass on ${ShoppingCart.TestCartClass}"))
+      entity.expectClosed()
+    }
+
+    "fail when snapshot handler throws exception" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart", snapshot(42, cartSnapshot())))
+      entity.expect(failure("Unexpected failure: Boom: no items"))
+      entity.expectClosed()
+    }
+
+    "fail when event handler does not exist" in {
+      val entity = client.connect
+      val notEvent = domainLineItem("?", "not an event", 1)
+      val eventClass = notEvent.getClass
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(event(1, notEvent))
+      entity.expect(failure(s"No event handler found for event $eventClass on ${ShoppingCart.TestCartClass}"))
+      entity.expectClosed()
+    }
+
+    "fail when event handler throws exception" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(event(1, itemAdded("123", "FAIL", 42)))
+      entity.expect(failure("Unexpected failure: Boom: name is FAIL"))
+      entity.expectClosed()
+    }
+
+    "fail when command handler does not exist" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(command(1, "cart", "foo"))
+      entity.expect(failure(1, s"No command handler found for command [foo] on ${ShoppingCart.TestCartClass}"))
+      entity.expectClosed()
+    }
+
+    "fail action when command handler uses context fail" in {
+      val entity = client.connect
+      entity.send(init(ShoppingCart.Name, "cart"))
+      entity.send(command(1, "cart", "AddItem", addItem("foo", "bar", -1)))
+      entity.expect(actionFailure(1, "Cannot add negative quantity of item [foo]"))
+      entity.passivate()
```

Comment:
Does the proxy passivate an entity on the reception of an action failure? If, yes, this would answer a question of mine regarding the user entities state, after such a failure.

Reply:
The proxy sees client action failures as an expected reply with a deliberate failure. It processes it the same as non-failed replies (even including persisting events and snapshots if they are included). The entity is not passivated, and the stream is not closed, for action failures. The examples use context.fail for validation, and I assume the expectation is for the entity to remain in a good state.

The `passivate` here in this test is just to end the interaction, while other tests have `expectClosed` if it's an entity failure plus stream closed on the service side — this is testing the java-support, while the proxy behaviour is tested in the exception handling spec.

Reply:
Once we're clear on the expected behaviour, we should express all of this in the TCK tests as well, and the specification.

Reply:
I agree